src/main/java/org/broadinstitute/hellbender/tools/evoquer/GnarlyGenotyperEngine.java

```diff
@@ -0,0 +1,492 @@
+        GnarlyGenotyperEngine.summarizePls = summarizePls;
+        GnarlyGenotyperEngine.keepAllSites = keepAllSites;
+        GnarlyGenotyperEngine.stripASAnnotations = stripASAnnotations;
+
+        if (!summarizePls) {
+            GenotypeLikelihoodCalculators GLCprovider = new GenotypeLikelihoodCalculators();
+
+            //initialize PL size cache -- HTSJDK cache only goes up to 4 alts, but I need 6
+            likelihoodSizeCache = new int[maxAltAllelesToOutput + 1];
+            for (final int numAlleles : IntStream.rangeClosed(1, maxAltAllelesToOutput + 1).boxed().collect(Collectors.toList())) {
+                likelihoodSizeCache[numAlleles - 1] = GenotypeLikelihoods.numLikelihoods(numAlleles, ASSUMED_PLOIDY);
+                glcCache.add(numAlleles - 1, GLCprovider.getInstance(ASSUMED_PLOIDY, numAlleles));
+            }
+        }
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public VariantContext finalizeGenotype(final VariantContext variant, final VariantContextBuilder annotationDBBuilder) {
+        //GenomicsDB or Evoquer merged all the annotations, but we still need to finalize MQ and QD annotations
+        //return a VC with the finalized annotations and dbBuilder gets the raw annotations for the database
+
+        final double QUALapprox = variant.getAttributeAsInt(GATKVCFConstants.RAW_QUAL_APPROX_KEY, 0);
+        //TODO: do we want to apply the indel prior to mixed sites?
+        final boolean isIndel = variant.getReference().length() > 1 || variant.getAlternateAlleles().stream().anyMatch(allele -> allele.length() > 1);
+        final double sitePrior = isIndel ? HomoSapiensConstants.INDEL_HETEROZYGOSITY : HomoSapiensConstants.SNP_HETEROZYGOSITY;
+        if((isIndel && QUALapprox < INDEL_QUAL_THRESHOLD) || (!isIndel && QUALapprox < SNP_QUAL_THRESHOLD)) {
+            if (keepAllSites) {
+                VariantContextBuilder builder = new VariantContextBuilder(mqCalculator.finalizeRawMQ(variant));
+                //we don't need the rest of the annotations since this is filtered and won't go to VQSR
+                builder.filter(GATKVCFConstants.LOW_QUAL_FILTER_NAME);
+                builder.attribute(GATKVCFConstants.AC_ADJUSTED_KEY, 0);
+                return builder.make();
+            }
+            return null;
+        }
+
+        //GenomicsDB merged all the annotations, but we still need to finalize MQ and QD annotations
+        //vcfBuilder gets the finalized annotations and annotationDBBuilder gets the raw annotations for the database
+        final VariantContextBuilder vcfBuilder = new VariantContextBuilder(mqCalculator.finalizeRawMQ(variant));
+
+        final int variantDP = variant.getAttributeAsInt(GATKVCFConstants.VARIANT_DEPTH_KEY, 0);
+        double QD = QUALapprox / (double)variantDP;
+        vcfBuilder.attribute(GATKVCFConstants.QUAL_BY_DEPTH_KEY, QD).log10PError(QUALapprox/-10.0-Math.log10(sitePrior));
```

Comment:
Can you provide a way to skip the QD calculation? (Had to be commented out for Evoquer). If you don't want to do it here, we can deal with it in the Evoquer PR instead.